guides/building-a-middleware-from-scratch.md

```diff
@@ -0,0 +1,693 @@
+means we can combine multiple errors type into one. That has the following
+advantages:
+
+1. Our error handling is less fragile since changing the order middleware are
+   applied in wont change the final error type.
+2. The error type now has a constant size regardless how many middleware we've
+   applied.
+3. Extracting the error no longer requires a big `match` but can instead be done
+   with `error.downcast_ref::<Timeout>()`.
+
+However it also has the following downsides:
+
+1. As we're using dynamic downcasting the compiler can no longer guarantee that
+   we're exhaustively checking for every possible error type.
+2. Creating an error now requires an allocation. In practice we expect errors to
+   be infrequent and therefore this shouldn't be a problem.
+
+Which option you prefer is a matter of personal preference. Both have their
+advantages and disadvantages. However the pattern that we've decided to use in
+Tower is boxed trait objects. You can find the original discussion
+[here](https://github.com/tower-rs/tower/issues/131).
+
+For our `Timeout` middleware that means we need to create a struct that
+implements `std::error::Error` such that we can convert it into a `Box<dyn
+std::error::Error + Send + Sync>`. We also have to require that the inner
+service's error type implements `Into<Box<dyn std::error::Error + Send +
+Sync>>`. Luckily most errors automatically satisfies that so it wont require
+users to write any additional code.
+
+The code for our error type looks like this:
+
+```rust
+use std::fmt;
+
+#[derive(Debug, Default)]
+pub struct TimeoutError(());
+
+impl fmt::Display for TimeoutError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.pad("request timed out")
+    }
+}
+
+impl std::error::Error for TimeoutError {}
+```
+
+We add a private field to `TimeoutError` such that users outside of Tower cannot
+construct their own `TimeoutError`. They can only be obtained through our
+middleware.
+
+`Box<dyn std::error::Error + Send + Sync>` is also quite a mouthful so
+lets define a type alias for it:
+
+```rust
+// This also exists as `tower::BoxError`
+pub type BoxError = Box<dyn std::error::Error + Send + Sync>;
+```
+
+Our future implementation now becomes:
+
+```rust
+impl<F, Response, Error> Future for ResponseFuture<F>
+where
+    F: Future<Output = Result<Response, Error>>,
+    // Require that the inner service's error can be converted into a `BoxError`.
+    Error: Into<BoxError>,
+{
+    type Output = Result<
+        Response,
+        // The error type of `ResponseFuture` is now `BoxError`.
+        BoxError,
+    >;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+
+        match this.response_future.poll(cx) {
+            Poll::Ready(result) => {
+                // Use `map_err` to convert the error type.
+                let result = result.map_err(Into::into);
```

Comment:
I _think_ you can just use `?` directly on `Poll<Result<_, _>>` now?

Reply:
Yeah I think you're right. I'm going to leave it as is because I think the additional verbosity might help those who haven't written futures before.